game/src/main/kotlin/world/gregs/voidps/world/activity/skill/herblore/PotionDecanting.kts

```diff
@@ -0,0 +1,114 @@
+    }
+}
+npcOperate("Talk-to", "bob_barter_herbs") {
+    npc<Cheerful>("Hello, chum, fancy buyin' some designer jewellry? They've come all the way from Ardougne! Most pukka!")
+    player<Talking>("Erm, no. I'm all set, thanks.")
+    npc<Cheerful>("Okay, chum, would you like to show you the very latest potion prices?")
+    choice {
+        option<Talk>("Who are you?") {
+            npc<Cheerful>("Why, I'm Bob! Your friendly seller of smashin' goods!")
+            player<Talking>("So what do you have to sell?")
+            npc<Talking>("Oh, not much at the moment. Cuz, ya know, business being so well and cushie.")
+            player<Talking>("You don't really look like you're being so successful.")
+            npc<Talking>("You plonka! It's all a show, innit! If I let people knows I'm in good business they'll want a share of the moolah!")
+            player<Talking>("You conveniently have a response for everything.")
+            npc<Talking>("That's the Ardougne way, my son.")
+        }
+        option<Talk>("Can you show me the prices for potions?") {
+            npc<Sad>("The Grand Exchange is still under construction, hopefully it won't take long, then i'll have those prices for yah")
+            // this should show an interface with price guide of herbs based off the grand exchange
+        }
+        option<Talk>("I'll leave you too it") { }
+    }
+}
+
+// no G.E so this cannot be implemented yet.
+npcOperate("Info-herbs", "bob_barter_herbs") {
+    return@npcOperate
+
+}
+
+fun decantPotions(player: Player): Boolean {
+    // Check if the player has any potions to decant
+    if (player.inventory.items.none { item -> potions.any { potion -> item.id.contains(potion) } }) {
+        return false
+    }
+
+    // Backup players inventory in case of an error we can restore the players items
+    val originalItems = player.inventory.items.toList()
+
+    try {
+        val potionMap = mutableMapOf<String, Int>()
+
+        // Remove potions and calculate total doses for each type
+        player.inventory.items.filter { item -> potions.any { potion -> item.id.contains(potion) } }
+                .toList().forEach { potion ->
+            val baseId = potion.id.substringBeforeLast('_')
+            potionMap[baseId] = potionMap.getOrDefault(baseId, 0) +
+                    Integer.parseInt(potion.id.takeLast(1)) * potion.amount
+            player.inventory.remove(potion.id)
+        }
+
+        // No Potion were decanted return false
+        if (potionMap.isEmpty()) {
+            return false
+        }
+
+        // Add decanted potions back into the players inventory
+        potionMap.forEach { (baseId, totalDoses) ->
+            var dosesLeft = totalDoses
+            while (dosesLeft > 0) {
+                val dose = if (dosesLeft >= 4) 4 else dosesLeft
+                dosesLeft -= dose
+                player.inventory.add("${baseId}_$dose", 1)
+            }
+        }
```

Comment:
The approach is fine but it would be better to use `transaction` so errors are handled in the standard way,
something along the lines of: (pseudo)
```kotlin
player.inventory.transaction {
    val potionMap = mutableMapOf()
    for (index in items.indices) {
        val item = items[index]
        if (isPotion(item)) {
            clear(index)
            val type = item.id.substringBeforeLast("_")
            val doses = item.id.substringAfterLast("_").toInt()
            potionMap[type] = potionMap.getOrDefault(type, 0) + doses
        }
    }
    for ((type, doses) in potionMap) {
        add("${type}_$dose", doses / 4)
        val remaining = doses.rem(4)
        if (remaining > 0) {
            add("${type}_${remaining}")
        }
    }
}
return player.inventory.transaction.error == TransactionError.None
```

Reply:
this code is a lot different from what I had written were you already working on decanting or did you just rewrite it? if you were already working on it sorry.

Reply:
I just rewrote what you had and simplified out the while loop 😄